File: 5 - Form Template Method/JS/CapitalStrategy.js
import RiskFactors from "./RiskFactors";
import UnusedRiskFactors from "./UnusedRiskFactors";

export default class CapitalStrategy {
    constructor() {
        this.millisPerDay = 86400000
        this.daysPerYear = 365
    }

    capital(loan) {
        throw new Error("implement in sub type")
    }

    riskFactor(loan) {
        return RiskFactors.ForRating(loan.getRiskRating())
    }

    unusedRiskFactor(loan){
        return UnusedRiskFactors.ForRating(loan.getRiskRating())
    }

    duration(loan){
        return this.yearsTo(loan.getExpiry(), loan)
    }

    yearsTo(endDate, loan) {
        const beginDate = loan.getToday() || loan.getStart()
        return Math.floor(((endDate - beginDate) / this.millisPerDay / this.daysPerYear))
    }
}

File: 5 - Form Template Method/C#/MPG.FormTemplateMethod.After/CapitalStrategy.cs
﻿namespace MPG.FormTemplateMethod.After
{
    using System;

    public abstract class CapitalStrategy
    {
        private const int MillisPerDay = 86400000;
        private const int DaysPerYear = 365;

        public virtual double Duration(Loan loan)
        {
            return YearsTo(loan.GetExpiry().Value, loan);
        }

        protected double RiskFactor(Loan loan)
        {
            return RiskFactors.ForRating(loan.GetRiskRating());
        }

        protected double UnusedRiskFactor(Loan loan)
        {
            return UnusedRiskFactors.ForRating(loan.GetRiskRating());
        }

        protected double YearsTo(DateTime endDate, Loan loan)
        {
            var beginDate = loan.GetToday().HasValue ? loan.GetToday().Value : loan.GetStart().Value;
            var beginDateMilliseconds = new DateTimeOffset(beginDate).ToUnixTimeMilliseconds();
            var endDateMilliseconds = new DateTimeOffset(endDate).ToUnixTimeMilliseconds();
            return ((endDateMilliseconds - beginDateMilliseconds) / MillisPerDay) / DaysPerYear;
        }

        protected virtual double RiskAmountFor(Loan loan)
        {
            return loan.GetCommitment();
        }

        public virtual double Capital(Loan loan)
        {
            return RiskAmountFor(loan) * Duration(loan) * RiskFactor(loan);
        }
    }
}